### ISSUE TITLE
Per-pixel focal loss

### ISSUE THREAD

**Issue by RubenS02** [jfzhang95/pytorch-deeplab-xception#133]
Should the focal loss not be applied per pixel? The reduction occurs when computing the crossentropy loss but shouldn't it occur at the very end?

    criterion = nn.CrossEntropyLoss(weight=self.weight, ignore_index=self.ignore_index,
                                        size_average=self.size_average)

    logpt = -criterion(logit, target.long())
    pt = torch.exp(logpt)
    loss = -((1 - pt) ** gamma) * logpt

should maybe become

    criterion = nn.CrossEntropyLoss(weight=self.weight, ignore_index=self.ignore_index,
                                        reduction="none") 
    # No reduction at the moment, the output is a tensor

    logpt = -criterion(logit, target.long())
    pt = torch.exp(logpt)
    loss = -((1 - pt) ** gamma) * logpt
    if self.size_average:
        loss = loss.mean()

**Comment by Rudychan** [comment]
> Should the focal loss not be applied per pixel? The reduction occurs when computing the crossentropy loss but shouldn't it occur at the very end?
> 
> ```
> criterion = nn.CrossEntropyLoss(weight=self.weight, ignore_index=self.ignore_index,
>                                     size_average=self.size_average)
> 
> logpt = -criterion(logit, target.long()) # logpt is a scalar
> pt = torch.exp(logpt)
> loss = -((1 - pt) ** gamma) * logpt
> ```
> 
> should maybe become
> 
> ```
> criterion = nn.CrossEntropyLoss(weight=self.weight, ignore_index=self.ignore_index,
>                                     reduction="none") 
> 
> logpt = -criterion(logit, target.long()) # logpt is a tensor
> pt = torch.exp(logpt)
> loss = -((1 - pt) ** gamma) * logpt
> if self.size_average:
>     loss = loss.mean()
> ```

I found the same problem, my solution is:
   `def FocalLoss(self, logit, target, gamma=2, alpha=0.5):
        n, c, h, w = logit.size()
        criterion = nn.CrossEntropyLoss(weight=self.weight, ignore_index=self.ignore_index,
                                        size_average=self.size_average)
        if self.cuda:
            criterion = criterion.cuda()

        logpt = -criterion(logit, target.long())
        pt = torch.exp(logpt)
        if alpha is not None:
            logpt *= alpha
        loss = -((1 - pt) ** gamma) * logpt

        return loss`

and  
 `print('Loss: %.3f' % train_loss*batch_size)`

**Comment by liuwenran** [comment]
agree with @RubenS02

**Comment by ycc66104116** [comment]
hi, i'm also curious about this issue. and want to know if this adjustment work successfully?

**Comment by RubenS02** [comment]
Hi @ycc66104116 , i did not test it, this repo is obsolete now you can find much better models
